ctr/exec-3dsx/exec_cia.c

```diff
@@ -0,0 +1,229 @@
+			return 1;
+	}
+	
+	return 0;
+}
+
+static int installCia(Handle ciaFile){
+	Result failed;
+	Handle outputHandle;
+	u64 fileSize;
+	u64 fileOffset = 0;
+	u32 bytesRead;
+	u32 bytesWritten;
+	u8 transferBuffer[FILE_CHUNK_SIZE];
+	
+	failed = AM_StartCiaInstall(MEDIATYPE_SD, &outputHandle);
+	if(R_FAILED(failed))
+		return -1;
+	
+	failed = FSFILE_GetSize(ciaFile, &fileSize);
+	if(R_FAILED(failed))
+		return -1;
+	
+	while(fileOffset < fileSize){
+		u64 bytesRemaining = fileSize - fileOffset;
+		failed = FSFILE_Read(ciaFile, &bytesRead, fileOffset, transferBuffer, bytesRemaining < FILE_CHUNK_SIZE ? bytesRemaining : FILE_CHUNK_SIZE);
+		if(R_FAILED(failed)){
+			AM_CancelCIAInstall(outputHandle);
+			return -1;
+		}
+		
+		failed = FSFILE_Write(outputHandle, &bytesWritten, fileOffset, transferBuffer, bytesRead, 0);
+		if(R_FAILED(failed)){
+			AM_CancelCIAInstall(outputHandle);
+			if(R_DESCRIPTION(failed) == RD_ALREADY_EXISTS)
+				return 1;
+			return -1;
+		}
+		
+		if(bytesWritten != bytesRead){
+			AM_CancelCIAInstall(outputHandle);
+			return -1;
+		}
+		
+		fileOffset += bytesWritten;
+	}
+	
+	failed = AM_FinishCiaInstall(outputHandle);
+	if(R_FAILED(failed))
+		return -1;
+	
+	return 1;
+}
+
+int exec_cia(const char* path, const char* args){
+	struct stat sBuff; 
+	bool fileExists;
+	bool inited;
+	
+	if(path == NULL || path[0] == '\0'){
+		errno = EINVAL;
+		return -1;
+	}
+	
+	fileExists = stat(path, &sBuff) == 0;
+	if(!fileExists){
+		errno = ENOENT;
+		return -1;
+	}
+	else if(S_ISDIR(sBuff.st_mode)){
+		errno = EINVAL;;
+		return -1;
+	}
+	
+	inited = R_SUCCEEDED(amInit()) && R_SUCCEEDED(fsInit());
+	if(inited){
+		Result res;
+		AM_TitleEntry ciaInfo;
+		FS_Archive ciaArchive;
+		Handle ciaFile;
+		int ciaInstalled;
+		ciaParam param;
+		int argsLength;
+		
+		//open cia file
+		res = FSUSER_OpenArchive(&ciaArchive, ARCHIVE_SDMC, fsMakePath(PATH_EMPTY, ""));
+		if(R_FAILED(res))
+			errorAndQuit("Cant open SD FS archive.");
+		
+		res = FSUSER_OpenFile(&ciaFile, ciaArchive, fsMakePath(PATH_ASCII, path + 5/*skip "sdmc:"*/), FS_OPEN_READ, 0);
+		if(R_FAILED(res))
+			errorAndQuit("Cant open CIA file.");
+	
+		res = AM_GetCiaFileInfo(MEDIATYPE_SD, &ciaInfo, ciaFile);
+		if(R_FAILED(res))
+			errorAndQuit("Cant get CIA file info.");
+		
+		ciaInstalled = isCiaInstalled(ciaInfo.titleID, ciaInfo.version);
+		if(ciaInstalled == -1){
+			//error
+			errorAndQuit("Could not read title id list.");
+		}
+		else if(ciaInstalled == 0){
+			//not installed
+			int error = installCia(ciaFile);
+			if(error == -1)
+				errorAndQuit("Cant install CIA.");
+		}
+		
+		FSFILE_Close(ciaFile);
+		FSUSER_CloseArchive(ciaArchive);
+		
+		param.argc = 0;
+		argsLength = 0;
+		if(args != NULL && args[0] != '\0'){
+			bool inSingleQuotes = false;
+			bool inDoubleQuotes = false;
+			int argStringLength = strlen(args);
+			
+			//build argument list like a terminal command on linux
+			for(unsigned int argPtr = 0; argPtr <= argStringLength; argPtr++){
```

Comment:
why this change ? there was already a simpler system in place, put argc first then the null terminated arguments after each other in the argv buffer, which is parsed on entry.

Reply:
Because I wanted it to be like a terminal on Linux so it can be used in other projects.

Reply:
??
this is C though not a Linux terminal.
you only need quotes on terminals because you can't input the null character.
you're complicating things by having to add quotes and then having to remove them for no gain.

Reply:
In c you can't use the null character either or it will terminate the string, and I wanted a single string argument list.

Reply:
that is why you put argc first in the buffer ....
the buffer has a fixed size so it will always be copied full.

Reply:
ok I assumed you were copying the buffer with quotes before launch and parsing them on entry.
you're not doing that , so you can get rid of the quote thing then no ? it makes no sense do to something then then directly call a function that reverts it.

Reply:
if you want your lib to be able to handle quotes strings as well, add a wrapper for that, but call the simpler function directly from RA.

Reply:
I can do that.

Reply:
I am sorry, its actually way to much extra work, and for no benefit, that isnt a speed intensive part of the program, its fine if it does one unnecessary operation.

It just makes the library more messy to add another argument format when this is what most people would want it to be anyway.

Reply:
this isn't a debate. the way it is right now is not acceptable.

Reply:
Its a nitpick, the library is meant to be it own thing, I am not changing it for this one edge case, if you have to don merge it, every time I try to submit something this happens over something irrelevant, its quotes or leave it crashing on launch like before, I am done wasting my time on this nonsense.